Docs: Migrate to react-router v6 (#718)
grafana/plugin-tools: docusaurus/docs/migration-guides/update-from-grafana-versions/v9.x-v10.x.md

Reviewer: We'll need to reference this in the Grafana docs as well as part of the deprecation notice for v5. My worry is that by adding here, given plugin devs may feel that they have already updated for v10 compatibility, it won't be the most discoverable location. That said - I still think it is the _correct_ location, so we just need to provide appropriate pointers.

Reviewer: raised issue to track

Author: Hmm, I think that these don't exist anymore inside the Grafana docs (they got removed in [this PR](https://github.com/grafana/grafana/pull/75325) if I am not mistaken). Is this what you meant?

Reviewer: No - https://grafana.com/docs/grafana/latest/breaking-changes/breaking-changes-v10-0/#for-plugin-developers

Author: Resolving this now as it is tracked in a separate issue.

When writing plugins that should run on 9.x, continue to use the Vector interfaces. In this case, when targeting versions 10+, you can now use simple arrays rather than wrapper classes.

To make this transition seamless, we employed the Original JavaScript Sin™. That is, we [extended the native Array prototype](https://github.com/grafana/grafana/blob/v10.0.x/packages/grafana-data/src/types/vector.ts) with several Vector methods. We will atone and undo this in v11, when Vector interfaces and classes are removed.

## Update to React Router v6

> Starting from Grafana 10 plugins should start updating to use the v6 version of `react-router`. Overall, `react-router` v6 aims to simplify route configuration and provide a more flexible and intuitive API for developers.

Reviewer 2: ```suggestion
Starting from Grafana 10, plugins should use the v6 of `react-router`. Overall, `react-router` v6 aims to simplify route configuration and provide a more flexible and intuitive API for developers.
```

> Starting from Grafana 10 plugins should start updating to use the v6 version of `react-router`. Overall, `react-router` v6 aims to simplify route configuration and provide a more flexible and intuitive API for developers.

Reviewer: ```suggestion
Starting from Grafana 10, plugins can start updating to use v6 of `react-router`. Overall, `react-router` v6 aims to simplify route configuration and provide a more flexible and intuitive API for developers.
```

Author: Fixed in https://github.com/grafana/plugin-tools/pull/718/commits/3d1b16ea721a02cd714cafaf8099390a8b220e25.

> **We strongly encourage developers to update their plugins to use the v6 version `react-router` as soon as possible, as the v5 version is going to be deprecated in the future.**

Reviewer: ```suggestion
If your current plugin version needs to maintain compatibility with Grafana v9, then you can continue to use `react-router@v5` in Grafana v10. Both versions are available for plugins. However, **we strongly encourage developers to update their plugins to use the v6 version `react-router` as soon as possible**, as the v5 version is going to be deprecated in Grafana v11 and subsequently removed.
```

wdyt?

Author: I think it's great 👍

[Official React Router v5 to v6 migration guide](https://reactrouter.com/en/main/upgrading/v5)

### Update using `@grafana/create-plugin`

> Follow the steps below to start using React router v6 in your plugin.

Reviewer 2: ```suggestion
Follow the steps below to start using React router v6 in your plugin:
```

> Follow the steps below to start using React router v6 in your plugin.

Reviewer: ```suggestion
Follow the steps below to start using `react-router` v6 in your plugin.
```
nit: consistency

Author: Nice catch!

> #### **1. Update the build related configuration:**

Reviewer 2: Can we omit the bold? If the default style doesn't look right, let me know and we can consider updating the CSS.

Author: Sure thing, thanks for spotting this @josmperez 👍

Author: Fixed in https://github.com/grafana/plugin-tools/pull/718/commits/dd37ffb7a3c3d18e886ff60e0e94790e088d786a.

Enable using `react-router@v6` by setting the following feature flag in `<project-root>/.cprc.json`:

```json
{
  "features": {
    "useReactRouterV6": true
  }
}
```

Now update the build configuration using the create-plugin tool:

```bash
npx @grafana/create-plugin@latest update
```

> After updating the build config you probably also need to update some things in your code. Please follow the steps below:

Reviewer 2: ```suggestion
After updating the build configuration, you probably also need to make some updates. Follow the steps below:
```

> After updating the build config you probably also need to update some things in your code. Please follow the steps below:

Reviewer: ```suggestion
After updating the build configuration, it is likely that you will need to make additional updates to your plugin. Please follow the steps below:
```

Author: I didn't update this one, as I already accepted Joe's suggestion for the same paragraph by the time I saw this. Can you please check if we should modify the current version? 🙏

Reviewer: I prefer ^ as it reads nicer to me and is explicit about updating the plugin logic, but I don't consider it a required change. Your call

Author: Applied in https://github.com/grafana/plugin-tools/pull/718/commits/250810c416281d0a0e025c56ac6b90e860642120.

#### **2. Use `<Routes>` instead of `<Switch>`**

```typescript
// Using <Routes> instead of <Switch> in `react-router` v6
import { Routes } from 'react-router-dom';

// ...

return (
  <Routes>
    <Route path="/" element={<Home />} />
  </Routes>
);
```

#### **3. Remove the `exact` prop from `<Route>` components**

```typescript
return (
  <Routes>
    {/* BAD (Until v5) */}
    <Route exact path="/" element={<Home />} />

    {/* GOOD (From v6) */}
    {/* (Routes are "exact" by default, you need to use the "*" to match sub-routes) */}
    <Route path="/" element={<Home />} />
  </Routes>
);

> ```

Reviewer 3: I'd add a step 4 here repeating the link to the migration guide, 

```
#### **4. Follow the migration guide for more in-depth change**
....
```

Author: Hmm, doesn't sound like a bad idea. Will update the PR. 👍

Author: Fixed in https://github.com/grafana/plugin-tools/pull/718/commits/21f6bd6ca0713734681a195146f827330cfb1f12.

## Update to React Router v6

Starting from Grafana 10, plugins can start using the v6 of `react-router`. Overall, `react-router` v6 aims to simplify route configuration and provide a more flexible and intuitive API for developers.

If your current plugin version needs to maintain compatibility with Grafana v9, then you can continue to use `react-router@v5` in Grafana v10. Both versions are available for plugins. However, **we strongly encourage developers to update their plugins to use the v6 version `react-router` as soon as possible**, as the v5 version is going to be deprecated in Grafana v11 and subsequently removed.

[Official React Router v5 to v6 migration guide](https://reactrouter.com/en/main/upgrading/v5)

### Update using `@grafana/create-plugin`

Follow the steps below to start using `react-router` v6 in your plugin:

#### 1. Update the build related configuration:

Enable using `react-router@v6` by setting the following feature flag in `<project-root>/.cprc.json`:

```json
{
  "features": {
    "useReactRouterV6": true
  }
}
```

Now update the build configuration using the create-plugin tool:

> ```bash

Reviewer: Ideally switch this to also include yarn/pnpm - could copy logic from `migration-guides/update-create-plugin-versions`

Author: Good call! Yeah, I will turn it into an `.mdx` and use the `<CodeSnippet>`.

When writing plugins that should run on 9.x, continue to use the Vector interfaces. In this case, when targeting versions 10+, you can now use simple arrays rather than wrapper classes.

To make this transition seamless, we employed the Original JavaScript Sin™. That is, we [extended the native Array prototype](https://github.com/grafana/grafana/blob/v10.0.x/packages/grafana-data/src/types/vector.ts) with several Vector methods. We will atone and undo this in v11, when Vector interfaces and classes are removed.

## Update to React Router v6

Starting from Grafana 10, plugins can start using the v6 of `react-router`. Overall, `react-router` v6 aims to simplify route configuration and provide a more flexible and intuitive API for developers.

If your current plugin version needs to maintain compatibility with Grafana v9, then you can continue to use `react-router@v5` in Grafana v10. Both versions are available for plugins. However, **we strongly encourage developers to update their plugins to use the v6 version `react-router` as soon as possible**, as the v5 version is going to be deprecated in Grafana v11 and subsequently removed.

> [Official React Router v5 to v6 migration guide](https://reactrouter.com/en/main/upgrading/v5)

Reviewer 2: ```suggestion
For more general information, refer to the [Official React Router v5 to v6 migration guide](https://reactrouter.com/en/main/upgrading/v5).
```

### Update using `@grafana/create-plugin`

Follow the steps below to start using `react-router` v6 in your plugin:

#### 1. Update the build related configuration:

Enable using `react-router@v6` by setting the following feature flag in `<project-root>/.cprc.json`:

```json
{
  "features": {
    "useReactRouterV6": true
  }
}
```

Now update the build configuration using the create-plugin tool:

<CodeSnippets
snippets={[
{ component: UpdateNPM, label: 'npm' },
{ component: UpdatePNPM, label: 'pnpm' },
{ component: UpdateYarn, label: 'yarn' },
]}
groupId="package-manager"
queryString="current-package-manager"
/>

> After updating the build configuration, it is likely that you will need to make additional updates to your plugin. Please follow the steps below:

Reviewer 2: ```suggestion
After updating the build configuration, it is likely that you will need to make additional updates to your plugin. To do so, follow the steps below:
```
